iOS/ViewControllers/UpcomingMoviesViewController.swift

```diff
@@ -0,0 +1,98 @@
+import UIKit
+import Shared
+
+protocol ListUpcomingMoviesDelegate: class {
+
+    func didList(movies: [Movie])
+    func didLoad(genres: [Genre])
+    func didDisplayError(message: String)
+    func didGetAvatar(movie: Movie, image: UIImage)
+
+}
+
+class UpcomingMoviesViewController: UIViewController, ListUpcomingMoviesDelegate {
+
+    private var cellIdentifier = String(describing: MovieTableViewCell.self)
+    private var tableViewDataSource: GenericTableViewDataSource<Movie, MovieTableViewCell>?
+    private var tableViewDelegate: GenericTableViewDelegate?
+    private var listUpcomingMoviesInteractor: ListUpcomingMoviesInteractor?
+    private var getMovieBackdropInteractor: GetMovieBackdropInteractor?
+    private var genres: [Genre] = []
+
+    @IBOutlet weak var tableView: UITableView!
+
+    override func viewDidLoad() {
+        configureGetMovieBackdropInteractor()
+        configureTableView()
+        configureUpcomingMoviesInteractor()
+        super.viewDidLoad()
+    }
+
+    private func configureGetMovieBackdropInteractor() {
+        getMovieBackdropInteractor = GetMovieBackdropInteractorFactory.make(
+            presenter: GetMovieBackdropPresenteriOS(delegate: self))
+    }
+
+    private func configureTableView() {
+        let cellNib = UINib(nibName: cellIdentifier, bundle: nil)
+        tableView.register(cellNib, forCellReuseIdentifier: cellIdentifier)
+        configureTableViewDataSource()
+        configureTableViewDelegate()
+    }
+
+    private func configureTableViewDataSource() {
+        tableViewDataSource = GenericTableViewDataSource() { (movie, cell) in
+            cell.configure(movie: movie)
+            let movieGenres = self.genres.filter({ movie.genresIds.contains($0.id) })
+            cell.configure(genres: movieGenres)
+            guard let dataSource = self.tableViewDataSource, let image = dataSource.images[movie.id] else {
+                self.getMovieBackdropInteractor?.get(movie: movie)
+                return
+            }
+            cell.configure(image: image)
+        }
+        tableView.dataSource = tableViewDataSource
+    }
+
+    private func configureTableViewDelegate() {
+        tableViewDelegate = GenericTableViewDelegate(selectedRow: { row in
+            if let movie = self.tableViewDataSource?.objects[row],
+                let navigationController = self.navigationController {
+                let movieImage = self.tableViewDataSource?.images[movie.id]
+                let movieGenres = self.genres.filter({ movie.genresIds.contains($0.id) })
+                MoviesRouterNavigation(navigationController: navigationController)
+                    .movieDetail(movie: movie, movieGenres: movieGenres, movieBackdropImage: movieImage)
+            }
+        })
+        tableView.delegate = tableViewDelegate
+    }
+
+    private func configureUpcomingMoviesInteractor() {
```

Comment:
Don't you think all of those configure could be done inside a Configurator or something like this ?
I know that you have already a Factory to create the Interactor, Presenter an so on, to help on DI. Just a thought